### ISSUE TITLE
A frame for Command Message is not found in binary protocol specification

### ISSUE THREAD

**Issue by t2y** [yahoo/pulsar#428]
According to [Framing](https://github.com/yahoo/pulsar/blob/master/docs/BinaryProtocol.md#framing), there're only 2 types of commands.

> * Simple commands
> [TOTAL_SIZE] [CMD_SIZE] [CMD]
> 
> * Payload commands
> [TOTAL_SIZE] [CMD_SIZE][CMD] [MAGIC_NUMBER][CHECKSUM] [METADATA_SIZE][METADATA] [PAYLOAD]

I found another frame of commands when a broker pushes Command Message to a client after the broker received Flow Command in [consumer](https://github.com/yahoo/pulsar/blob/master/docs/BinaryProtocol.md#consumer) process.

In fact, Command Message uses another frame format as below. It seems that *[MAGIC_NUMBER][CHECKSUM]* is removed as compared to *Payload commands*.

```java
private static ByteBuf serializeCommandMessageWithSize(BaseCommand cmd, ByteBuf metadataAndPayload) {
    // / Wire format
    // [TOTAL_SIZE] [CMD_SIZE][CMD] [METADATA_SIZE][METADATA] [PAYLOAD]
    ...
```

* https://github.com/yahoo/pulsar/blob/7badf1a/pulsar-common/src/main/java/com/yahoo/pulsar/common/api/Commands.java#L735

#### Expected behavior

```
[TOTAL_SIZE] [CMD_SIZE][CMD] [METADATA_SIZE][METADATA] [PAYLOAD]
```

The spec document should have this frame and explains related matters.

#### Actual behavior

There're no documents.

**Comment by merlimat** [comment]
Hi @t2y, the comment in the code is actually wrong, or better said outdated.

Typically, the `metadataAndPayload` that is passed, starts with the `[MAGIC_NUMBER][CHECKSUM] ...` header. 

The only reason why it might not be the case is if the client library is very old (in fact, that can only be true at Yahoo, since the checksum was added before open-sourcing the code).

So, if the producer included the checksum, but the consumer was not updated yet, the broker would strip it before sending.  https://github.com/yahoo/pulsar/blob/a2ce53608916eb60a88d7a57c69e627666dbdb4e/pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/Consumer.java#L184

Right now, I think we should just update the comment to reflect the right wire format that is applied.

**Comment by t2y** [comment]
Hi @merlimat , thank you for explaining the background. I think I could understand what the pulsar producer and consumer currently behave.

Let me confirm my understanding.

My debugging environment is here. I started standalone with almost default settings.

```bash
$ bin/pulsar standalone
```

I use java pulsar-client and verify communication between producer and consumer as below.

```bash
$ bin/pulsar-client produce -m 'hello' 'persistent://sample/standalone/ns1/my-topic'
```
then,
```bash
$ bin/pulsar-client consume -s 'sub' 'persistent://sample/standalone/ns1/my-topic'
```

I learned that `serializeCommandMessageWithSize()` method just handles `metadataAndPayload`, and it can include the checksum if a producer stores it (into bookkeeper's entry?). But, in my environment, a broker send Command Message without checksum when a consumer would request to retrieve.

Then, I found pulsar-client also can handles verifying `[MAGIC_NUMBER]`.

```java
    void messageReceived(MessageIdData messageId, ByteBuf headersAndPayload, ClientCnx cnx) {
    ...
        if (!verifyChecksum(headersAndPayload, messageId)) {
            // discard message with checksum error
            discardCorruptedMessage(messageId, cnx, ValidationError.ChecksumMismatch);
            return;
        }
```

https://github.com/yahoo/pulsar/blob/7badf1a/pulsar-client/src/main/java/com/yahoo/pulsar/client/impl/ConsumerImpl.java#L672

That is to say, it seems that broker and client can handle both the message with checksum and the message without checksum. Probably, in the future, a broker will handle only the message with checksum.

I can understand this behavior is not specification, but it is just happened in transition phase.

> Right now, I think we should just update the comment to reflect the right wire format that is applied.

I agree with it.
